presto-main/src/main/java/com/facebook/presto/server/SimpleHttpResponseHandler.java

```diff
@@ -57,9 +58,14 @@ else if (response.getStatusCode() == HttpStatus.SERVICE_UNAVAILABLE.code()) {
                 callback.failed(new ServiceUnavailableException(uri));
             }
             else {
-                // Something is broken in the server or the client, so fail immediately (includes 500 errors)
                 Exception cause = response.getException();
-                if (cause == null) {
+                String responseErrorCode = response.getHeader("Error-Code");
+                if (responseErrorCode != null) {
+                    StandardErrorCode code = StandardErrorCode.valueOf(responseErrorCode);
```

Comment:
There is a chance that the error code is not from `StandardErrorCode` (mostly because of it is a connector error code). Then this will throw an exception as the main engine has no knowledge of connectors. My suggestion is to try/catch that specific error and throw with an INTERNAL error as a catch-all.

Reply:
The error code on the worker side is set within presto error code scope. We have a contract that the error code contains only presto error code within StandardErrorCode file.